mithril-aggregator/src/snapshotter.rs

```diff
@@ -37,6 +44,64 @@ impl From<ZstandardCompressionParameters> for SnapshotterCompressionAlgorithm {
     }
 }
 
+/// Define multiple ways to append content to a tar archive.
+trait TarAppender {
+    fn append<T: Write>(&self, tar: &mut tar::Builder<T>) -> StdResult<()>;
+}
+
+struct AppenderDirAll {
+    db_directory: PathBuf,
+}
+
+impl TarAppender for AppenderDirAll {
+    fn append<T: Write>(&self, tar: &mut tar::Builder<T>) -> StdResult<()> {
+        tar.append_dir_all(".", &self.db_directory)
+            .map_err(SnapshotError::CreateArchiveError)
+            .with_context(|| {
+                format!(
+                    "Can not add directory: '{}' to the archive",
+                    self.db_directory.display()
+                )
+            })?;
+        Ok(())
+    }
+}
+
+struct AppenderEntries {
+    entries: Vec<PathBuf>,
+    db_directory: PathBuf,
+}
+
+impl TarAppender for AppenderEntries {
+    fn append<T: Write>(&self, tar: &mut tar::Builder<T>) -> StdResult<()> {
+        for entry in &self.entries {
+            let entry_path = self.db_directory.join(entry);
+            if entry_path.is_dir() {
+                tar.append_dir_all(entry, entry_path.clone())
+                    .with_context(|| {
+                        format!(
+                            "Can not add directory: '{}' to the archive",
+                            entry_path.display()
+                        )
+                    })?;
+            } else if entry_path.is_file() {
+                let mut file = File::open(entry_path.clone())?;
+                tar.append_file(entry, &mut file).with_context(|| {
+                    format!(
+                        "Can not add file: '{}' to the archive",
+                        entry_path.display()
+                    )
+                })?;
+            } else {
+                return Err(anyhow!(
```

Comment:
What would it be in that case? Do we want to raise an error or skip?

Reply:
It seemed more coherent to us to leave the responsibility for handling errors to the caller.

Reply:
So if any of the entries is neither a file nor a directory (then what would it be?), the archive creation is doomed.
Is this really the behavior that we expect?

Reply:
The error message has been reworded to be more relevant: https://github.com/input-output-hk/mithril/pull/2189/files#diff-54bfe4d87d887f1cae0f5ce311f07e0bfb726b285085fbea640d9689c48149c4R97